Sources/Contour/ConnectionPool.cs

```diff
@@ -0,0 +1,179 @@
+﻿namespace Contour
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using Common.Logging;
+
+    internal abstract class ConnectionPool<TConnection> : IConnectionPool<TConnection> where TConnection : class, IConnection
+    {
+        private readonly object syncRoot = new object();
+        private readonly ILog logger = LogManager.GetLogger<ConnectionPool<TConnection>>();
+        private readonly ConcurrentDictionary<string, IList<Tuple<TConnection, bool>>> groups =
+            new ConcurrentDictionary<string, IList<Tuple<TConnection, bool>>>();
+
+        private bool disposed;
+        private CancellationTokenSource cancellation;
+        
+        protected ConnectionPool()
+        {
+            this.cancellation = new CancellationTokenSource();
+        }
+
+        public event EventHandler ConnectionOpened;
+
+        public event EventHandler ConnectionClosed;
+
+        public event EventHandler ConnectionDisposed;
+        
+        public int Count => this.groups.SelectMany(pair => pair.Value).Count();
+
+        protected IConnectionProvider<TConnection> Provider { get; set; }
+
+        /// <summary>
+        /// Gets a new connection from the pool or uses an existing one
+        /// </summary>
+        /// <param name="connectionString">A connection string to create a new or get an existing connection</param>
+        /// <param name="reusable">Specifies if a connection can be reused</param>
+        /// <param name="token">Operation cancellation token</param>
+        /// <returns>A pooled connection</returns>
+        public TConnection Get(string connectionString, bool reusable, CancellationToken token)
+        {
+            lock (this.syncRoot)
+            {
+                if (this.disposed)
+                {
+                    throw new ObjectDisposedException(nameof(ConnectionPool<TConnection>));
```

Comment:
Just to specify the name (the name of the type actually) of the object being disposed.